src/arch/x86_64/CodeGen.zig

```diff
@@ -349,6 +349,13 @@ pub fn addExtraAssumeCapacity(self: *Self, extra: anytype) u32 {
 fn gen(self: *Self) InnerError!void {
     const cc = self.fn_type.fnCallingConvention();
     if (cc != .Naked) {
+        // push the callee_preserved_regs that were used
+        const backpatch_push_callee_preserved_regs_i = try self.addInst(.{
```

Comment:
This might be obvious but why do we need a special pseudo-instruction for register allocation?

Reply:
This is not register allocation, but pushing/popping registers. The reason that it is necessary is that before we gen the body, we don't know which registers are used. So we need to backpatch them. The reason why multiple push instructions can't be done is that we don't know which registers need to be pushed, and therefore we don't know how many mir indexes it will take.

Reply:
Oh OK, I think I get it now. So we basically need a placeholder within MIR to be able to say that in this place in code we will need to preserve a couple registers and then pop them in reverse order but we don't yet know which ones until we finish generating codegen for the given block / scope, correct? Hmm, there might not be a better way to do this in one pass than what you suggested here, but I'm wondering if it was worth considering an alternative: for every block / scope that codegen needs to generate, we would open a new "subslice" of MIR instructions which we would then be free to fixup with the correct number of push / pop instructions and only then we would append the entire thing to the global `mir_inststructions` multi arraylist. I think this approach might be worth considering since it would remove the need for a compound MIR instruction such as `push_regs_from_callee_preserved_regs`, and the resulting MIR slice would only consist of granual push / pop instructions. I haven't given it that much thought yet but this might be a slightly better input for any subsequent optimisation passes we'll want to perfom on the codegen in the future perhaps? @g-w1 @joachimschmidt557 lemme know what you guys think about this proposal.

Reply:
The point about further optimization passes makes a lot of sense. I don't really understand how the "subslice" would work with prepending it to the function's Mir.

Reply:
Oh, sorry @g-w1, that's me being a poor communicator. What I meant was that we need to do this backpatching any time we sort of open a new scope. Then, we would simply accumulate the MIR instructions local to this scope in a temp multi arraylist (or similar), which we would then fixup using your logic while leaving the scope, and we would append the fixed-up list to the global `mir_instructions`. Something along the lines of:

```zig
var local_mir = std.MultiArrayList(Mir).init(gpa);
defer local_mir.deinit();

// do all the logic here like the logic from gen() or similar

// since local_mir is scoped to this local block, we can now backpatch push/pops granularly like

inline for (...) |reg| {
    try self.addInst(push(reg));
}

try self.mir_instructions.appendSlice(local_mir);

inline for (...) |reg| {
    try self.addInst(pop(reg));
}
```

Lemme know if that makes sense! Sorry for the pseudo-zig but I'm on my phone ATM so trying to just get the message across ;-)

Reply:
Ahh, this totally makes sense. I'll switch to it.

Reply:
Nice! Try it out and lemme know how this goes. In the meantime, when I have some time to access my laptop, I'll help you in debugging the clobbering issue too.

Reply:
Introducing MIR sub-slices may introduce a new issue: In branches, we reference other MIR instructions by their absolute index. When a fresh MIR sub-slice needs to be appended to the global MIR list, we need to rewrite all of those references to the new, global indices. Alternatively, we could use relative offsets (which would then need to be signed integers to enable backwards jumps).

Reply:
Mhm, that's a good point. Having said that though, push and pop instructions in the scopes prologue and epilogue should be fairly standard and would not have a jump to any one instruction, right? That is, any jump reloc from within the block would not end up in those regions but rather either after the scope's prologue or before the epilogue, correct? Perhaps it'd be worth saving the relative indices of jmp instructions that require a reloc in a separate arraylist side-by-side with the MIR subslice, and after working out the push/pop prologue/epilogue just translating the jmp targets by the offset equivalent to the prologue length? I think this should be doable and fairly efficient, requiring surgical fixups of a couple MIR instructions only. However, if it becomes overly complicated, then I'm happy to revisit this in the future instead.

Reply:
Here is my diff for this feature 
```diff
diff --git a/lib/std/multi_array_list.zig b/lib/std/multi_array_list.zig
index fd387513b..7ec67785b 100644
--- a/lib/std/multi_array_list.zig
+++ b/lib/std/multi_array_list.zig
@@ -174,6 +174,18 @@ pub fn MultiArrayList(comptime S: type) type {
             self.set(self.len - 1, elem);
         }
 
+        /// Append the slice of items to the list. Allocates more
+        /// memory as necessary.
+        /// TODO this can be optimised for speed
+        pub fn appendSlice(self: *Self, gpa: *Allocator, s: Slice) !void {
+            try self.ensureUnusedCapacity(gpa, s.len);
+            var i: usize = 0;
+            const slice_marlist = s.toMultiArrayList();
+            while (i < s.len) : (i += 1) {
+                self.appendAssumeCapacity(slice_marlist.get(i));
+            }
+        }
+
         /// Extend the list by 1 element, asserting `self.capacity`
         /// is sufficient to hold an additional item.  Returns the
         /// newly reserved index with uninitialized data.
diff --git a/src/arch/x86_64/CodeGen.zig b/src/arch/x86_64/CodeGen.zig
index 831093a87..e1a7c2551 100644
--- a/src/arch/x86_64/CodeGen.zig
+++ b/src/arch/x86_64/CodeGen.zig
@@ -86,6 +86,10 @@ next_stack_offset: u32 = 0,
 /// Debug field, used to find bugs in the compiler.
 air_bookkeeping: @TypeOf(air_bookkeeping_init) = air_bookkeeping_init,
 
+/// How much we should add to any Data with `inst` tag set, to offset for
+/// instructions that we backpatch.
+add_to_jmp_reloc: u32 = undefined,
+
 const air_bookkeeping_init = if (std.debug.runtime_safety) @as(usize, 0) else {};
 
 pub const MCValue = union(enum) {
@@ -349,12 +353,10 @@ pub fn addExtraAssumeCapacity(self: *Self, extra: anytype) u32 {
 fn gen(self: *Self) InnerError!void {
     const cc = self.fn_type.fnCallingConvention();
     if (cc != .Naked) {
-        // push the callee_preserved_regs that were used
-        const backpatch_push_callee_preserved_regs_i = try self.addInst(.{
-            .tag = .push_regs_from_callee_preserved_regs,
-            .ops = undefined,
-            .data = .{ .regs_to_push_or_pop = undefined }, // to be backpatched
-        });
+        const global_instructions = self.mir_instructions;
+        var local_mir = std.MultiArrayList(Mir.Inst){};
+        defer local_mir.deinit(self.gpa);
+        self.mir_instructions = local_mir;
 
         _ = try self.addInst(.{
             .tag = .push,
@@ -430,22 +432,35 @@ fn gen(self: *Self) InnerError!void {
             }).encode(),
             .data = undefined,
         });
-        // calculate the data for callee_preserved_regs to be pushed and popped
-        var callee_preserved_regs_push_data: u32 = 0x0;
-        inline for (callee_preserved_regs) |reg, i| {
+        self.mir_instructions = global_instructions;
+        // push callee_preserved_regs that were used
+        inline for (callee_preserved_regs) |reg| {
             if (self.register_manager.isRegAllocated(reg)) {
-                callee_preserved_regs_push_data |= 1 << @intCast(u5, i);
+                _ = try self.addInst(.{
+                    .tag = .push,
+                    .ops = (Mir.Ops{
+                        .reg1 = reg,
+                    }).encode(),
+                    .data = undefined,
+                });
             }
         }
-        const data = self.mir_instructions.items(.data);
-        // backpatch the push instruction
-        data[backpatch_push_callee_preserved_regs_i].regs_to_push_or_pop = callee_preserved_regs_push_data;
-        // pop the callee_preserved_regs
-        _ = try self.addInst(.{
-            .tag = .pop_regs_from_callee_preserved_regs,
-            .ops = undefined,
-            .data = .{ .regs_to_push_or_pop = callee_preserved_regs_push_data },
-        });
+        self.add_to_jmp_reloc = @intCast(u32, global_instructions.len);
+        // add the body
+        try self.mir_instructions.appendSlice(self.gpa, local_mir.slice());
+        // pop them
+        inline for (callee_preserved_regs) |reg| {
+            if (self.register_manager.isRegAllocated(reg)) {
+                _ = try self.addInst(.{
+                    .tag = .pop,
+                    .ops = (Mir.Ops{
+                        .reg1 = reg,
+                    }).encode(),
+                    .data = undefined,
+                });
+            }
+        }
+
         _ = try self.addInst(.{
             .tag = .ret,
             .ops = (Mir.Ops{
diff --git a/src/arch/x86_64/Emit.zig b/src/arch/x86_64/Emit.zig
index 7fd81d46c..8f7995474 100644
--- a/src/arch/x86_64/Emit.zig
+++ b/src/arch/x86_64/Emit.zig
@@ -285,7 +285,7 @@ fn mirJmpCall(emit: *Emit, tag: Mir.Inst.Tag, inst: Mir.Inst.Index) InnerError!v
     const ops = Mir.Ops.decode(emit.mir.instructions.items(.ops)[inst]);
     const flag = @truncate(u1, ops.flags);
     if (flag == 0) {
-        const target = emit.mir.instructions.items(.data)[inst].inst;
+        const target = emit.mir.instructions.items(.data)[inst].inst + emit.mir.function.add_to_jmp_reloc;
         const opc: u8 = switch (tag) {
             .jmp => 0xe9,
             .call => 0xe8,
@@ -441,7 +441,7 @@ inline fn getCondOpCode(tag: Mir.Inst.Tag, cond: CondType) u8 {
 
 fn mirCondJmp(emit: *Emit, tag: Mir.Inst.Tag, inst: Mir.Inst.Index) InnerError!void {
     const ops = Mir.Ops.decode(emit.mir.instructions.items(.ops)[inst]);
-    const target = emit.mir.instructions.items(.data)[inst].inst;
+    const target = emit.mir.instructions.items(.data)[inst].inst + emit.mir.function.add_to_jmp_reloc;
     const cond = CondType.fromTagAndFlags(tag, ops.flags);
     const opc = getCondOpCode(tag, cond);
     const source = emit.code.items.len;
```
IMO, we should stick to the backpatch strategy for now, as this is pretty complicated (and still does not work for now).
Also, copying the local slice into the mir_instructions is O(n), which is not ideal.

Reply:
Mhm, sounds good to me! Let's stick to the backpatch strategy for now.